chunky/src/java/se/llbit/chunky/renderer/scene/PathTracer.java

```diff
@@ -123,270 +123,31 @@ public static boolean pathTrace(Scene scene, Ray ray, WorkerState state, int add
-                }
-
-                reflected.setCurrentMaterial(reflected.getPrevMaterial(), reflected.getPrevData());
-
-                getDirectLightAttenuation(scene, reflected, state);
-
-                Vector4 attenuation = state.attenuation;
-                if (attenuation.w > 0) {
-                  double mult = QuickMath.abs(reflected.d.dot(ray.getNormal())) * (scene.getSunSamplingStrategy().isSunLuminosity() ? scene.sun().getLuminosityPdf() : 1);
-                  directLightR = attenuation.x * attenuation.w * mult;
-                  directLightG = attenuation.y * attenuation.w * mult;
-                  directLightB = attenuation.z * attenuation.w * mult;
-                  hit = true;
-                }
-              }
-
-              reflected.diffuseReflection(ray, random);
-              hit = pathTrace(scene, reflected, state, 0, false) || hit;
-              if (hit) {
-                ray.color.x = ray.color.x * (emittance + directLightR * scene.sun.emittance.x + (
-                    reflected.color.x + reflected.emittance.x) + (indirectEmitterColor.x));
-                ray.color.y = ray.color.y * (emittance + directLightG * scene.sun.emittance.y + (
-                    reflected.color.y + reflected.emittance.y) + (indirectEmitterColor.y));
-                ray.color.z = ray.color.z * (emittance + directLightB * scene.sun.emittance.z + (
-                    reflected.color.z + reflected.emittance.z) + (indirectEmitterColor.z));
-              } else if(indirectEmitterColor.x > Ray.EPSILON || indirectEmitterColor.y > Ray.EPSILON || indirectEmitterColor.z > Ray.EPSILON) {
-                hit = true;
-                ray.color.x *= indirectEmitterColor.x;
-                ray.color.y *= indirectEmitterColor.y;
-                ray.color.z *= indirectEmitterColor.z;
-              }
-
-            } else {
-              reflected.diffuseReflection(ray, random);
-
-              hit = pathTrace(scene, reflected, state, 0, false) || hit;
-              if (hit) {
-                ray.color.x =
-                    ray.color.x * (emittance + (reflected.color.x + reflected.emittance.x) + (indirectEmitterColor.x));
-                ray.color.y =
-                    ray.color.y * (emittance + (reflected.color.y + reflected.emittance.y) + (indirectEmitterColor.y));
-                ray.color.z =
-                    ray.color.z * (emittance + (reflected.color.z + reflected.emittance.z) + (indirectEmitterColor.z));
-              } else if(indirectEmitterColor.x > Ray.EPSILON || indirectEmitterColor.y > Ray.EPSILON || indirectEmitterColor.z > Ray.EPSILON) {
-                hit = true;
-                ray.color.x *= indirectEmitterColor.x;
-                ray.color.y *= indirectEmitterColor.y;
-                ray.color.z *= indirectEmitterColor.z;
-              }
-            }
-          }
+      int count = firstReflection ? scene.getBranchCount() : 1;
+      for (int i = 0; i < count; i++) {
```

Comment:
A comment on this technique (is there a paper or some other source on this?) would be great.

Just an idea: Could we re-use more than the first ray? Eg. re-use entire ray paths _but_ the last ray?